external-crates/move/move-abstract-stack/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+    /// Creates an empty stack
+    pub fn new() -> Self {
+        Self {
+            values: vec![],
+            len: 0,
+        }
+    }
+
+    /// Returns true iff the stack is empty
+    pub fn is_empty(&self) -> bool {
+        // empty ==> len is 0
+        debug_assert!(!self.values.is_empty() || self.len == 0);
+        // !empty ==> last element len <= len
+        debug_assert!(self.values.is_empty() || self.values.last().unwrap().0 <= self.len);
+        self.values.is_empty()
+    }
+
+    /// Returns the logical length of the stack as if there was no space
+    /// compression
+    pub fn len(&self) -> u64 {
+        // len is 0 ==> empty
+        debug_assert!(self.len != 0 || self.values.is_empty());
+        // len not 0 ==> !empty and last element len <= len
+        debug_assert!(
+            self.len == 0 || (!self.values.is_empty() && self.values.last().unwrap().0 <= self.len)
+        );
+        self.len
+    }
+
+    /// Pushes a single value on the stack
+    pub fn push(&mut self, item: T) -> Result<(), AbsStackError> {
+        self.push_n(item, 1)
+    }
+
+    /// Push n copies of an item on the stack
+    pub fn push_n(&mut self, item: T, n: u64) -> Result<(), AbsStackError> {
+        if n == 0 {
+            return Ok(());
+        }
+
+        let Some(new_len) = self.len.checked_add(n) else {
+            return Err(AbsStackError::Overflow);
+        };
+        self.len = new_len;
+        match self.values.last_mut() {
+            Some((count, last_item)) if &item == last_item => {
+                debug_assert!(*count > 0);
+                *count += n
+            }
+            _ => self.values.push((n, item)),
+        }
+        Ok(())
+    }
+
+    /// Pops a single value off the stack
+    pub fn pop(&mut self) -> Result<T, AbsStackError> {
+        self.pop_eq_n(NonZeroU64::new(1).unwrap())
+    }
+
+    /// Pops n values single value off the stack, erroring if non enough items or if the n items are
+    /// not equal
+    pub fn pop_eq_n(&mut self, n: NonZeroU64) -> Result<T, AbsStackError> {
+        let n: u64 = n.get();
+        if self.is_empty() {
+            return Err(AbsStackError::EmptyStack);
+        }
+        if n > self.len {
+            return Err(AbsStackError::NotEnoughValues);
+        }
+        let (count, last) = self.values.last_mut().unwrap();
+        debug_assert!(*count > 0);
+        let ret = match (*count).cmp(&n) {
+            Ordering::Less => return Err(AbsStackError::ElementNotEqual),
+            Ordering::Equal => {
+                let (_, last) = self.values.pop().unwrap();
+                last
+            }
+            Ordering::Greater => {
+                *count -= n;
+                last.clone()
+            }
+        };
+        self.len -= n;
+        Ok(ret)
+    }
+
+    /// Pop any n items off the stack. Unlike `pop_n`, items do not have to be equal
+    pub fn pop_any_n(&mut self, n: NonZeroU64) -> Result<(), AbsStackError> {
```

Comment:
I agree as it would be really strange to leave it partially updated.
I changed it last commit to check for the length first and to error without mutations if there are not enough values